### PR TITLE
Add new "Add volume" button, to add volume to VM instance directly

### PR DESCRIPTION
**Pull request by kioie**
## Description

Volume button to allow user to quickly add volume to a VM instance.
Also slight edit of createVolumeCmd to allow for direct volume attachment to VM.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Screen Shot 2019-07-30 at 8 31 17 PM

Screen Shot 2019-07-30 at 8 35 08 PM

Screen Shot 2019-07-30 at 8 35 46 PM
Volume already attached to instance

Screen Shot 2019-07-30 at 8 37 38 PM
Trying to attach Volume to VM instance with snapshots
## How Has This Been Tested?

Tested createVolume API
Tried to attach Volume to VM instance with snapshots
Tried to attach Volume to VM instance without snapshots
Tried to attach Volume to running VM
Tried to attach Volume to stopped VM

### CODE REVIEW

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -736,8 +717,47 @@ public VolumeVO allocVolume(CreateVolumeCmd cmd) throws ResourceAllocationExcept
 
         String userSpecifiedName = getVolumeNameFromCommand(cmd);
 
-        return commitVolume(cmd, caller, owner, displayVolume, zoneId, diskOfferingId, provisioningType, size, minIops, maxIops, parentVolume, userSpecifiedName,
+        VolumeVO volume= commitVolume(cmd, caller, owner, displayVolume, zoneId, diskOfferingId, provisioningType, size,
+                minIops,
+                maxIops, parentVolume, userSpecifiedName,
                 _uuidMgr.generateUuid(Volume.class, cmd.getCustomId()));
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: @kioie can you modularise this code somewhat? as it has moved it is hard to compare with the old situation and the method is to complex for an overview at first glance.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -736,8 +717,47 @@ public VolumeVO allocVolume(CreateVolumeCmd cmd) throws ResourceAllocationExcept
...
+            }
+
+            // Check that the VM is in the correct state
+            if (vm.getState() != State.Running && vm.getState() != State.Stopped) {
+                throw new InvalidParameterValueException("Please specify a VM that is either running or stopped.");
+            }
+
+            // permission check
+            _accountMgr.checkAccess(caller, null, false, vm);
+            {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: why this extra block level?

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 3]
```diff
@@ -736,8 +717,47 @@ public VolumeVO allocVolume(CreateVolumeCmd cmd) throws ResourceAllocationExcept
...
+                try {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: maybe extract this try block as well into a separate method.?!

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 1 above)
Comment (kioie) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Is this better now?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: :) yes of course it is better, however...

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 4]
```diff
@@ -1662,7 +1629,45 @@ public Volume attachVolumeToVM(Long vmId, Long volumeId, Long deviceId) {
             return vol;
         }
     }
+    
+    public Volume verifyVmAndAttachVolume(Long vmId, VolumeVO volume){
 
+        Account caller = CallContext.current().getCallingAccount();
+        Volume vol = null;
+        if (vmId != null) {
+            // Check that the virtual machine ID is valid and it's a user vm
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: I think these comments warrant a further modularisation, so
checkVmIdIsValidUservm()

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 5]
```diff
@@ -1662,7 +1629,45 @@ public Volume attachVolumeToVM(Long vmId, Long volumeId, Long deviceId) {
...
+            UserVmVO vm = _userVmDao.findById(vmId);
+            if (vm == null || vm.getType() != VirtualMachine.Type.User) {
+                throw new InvalidParameterValueException("Please specify a valid User VM.");
+            }
+
+            // Check that the VM is in the correct state
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: checkVmIsInCorrectState()

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 6]
```diff
@@ -1662,7 +1629,45 @@ public Volume attachVolumeToVM(Long vmId, Long volumeId, Long deviceId) {
...
+            if (vm.getState() != State.Running && vm.getState() != State.Stopped) {
+                throw new InvalidParameterValueException("Please specify a VM that is either running or stopped.");
+            }
+
+            // permission check
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: checkPermissions()

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 4 above)
Comment (kioie) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: done

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 5 above)
Comment (kioie) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: done

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 6 above)
Comment (kioie) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: done...but I think, i'll edit this to make it abit more usable globally

Comment (kioie) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: done

**Comment by DaanHoogland** [comment]
@andrijapanicsb I put this to milestone 4.15, which means the ui part needs rewriting. (cc @rhtyd )
functional change but we can decide to sneak it into 4.14 as well

**Comment by rohityadavcloud** [comment]
Thanks for the PR @kioie - are you able to work on this, or is this work/PR abandoned?

**Comment by rohityadavcloud** [comment]
Hi @kioie, thanks for the PR. Due to inactivity on the PR and that the changes are in the legacy UI, the PR can't be unfortunately. If you wish to work/resume on this, pl submit a new PR with changes against the current UI.

**Pull request closed without merging**